---
title: "Ph-IRES phylogeny"
date: 2022-07-01
output: Newick-formatted tree file
---
```{r setup}
##finds if required packages are installed, loads them if not
if (!requireNamespace("BiocManager", quietly=TRUE)) install.packages("BiocManager")
if (!require("phangorn")) install.packages("phangorn")
if (!require("ape")) install.packages("ape")
if (!require("msa")) BiocManager::install("msa")
if (!require("tinytex")) install.packages("tinytex") 
if (!require("devtools")) install.packages("devtools") +
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("igraph")) install.packages("igraph", type="binary")
if (!require("ips")) install.packages("ips")
if(!require("seqinr")) install.packages("seqinr")
if(!require("ggtext")) install.packages("ggtext")
if(!require("bios2mds")) install.packages("bios2mds")
if(!require("Biostrings")) BiocManager::install("Biostrings")
if(!require("microRNA")) BiocManager::install("microRNA")
if(!require("ggtree")) BiocManager::install("ggtree")
if(!require("ggimage")) BiocManager::install("ggimage", type="binary")
if (!require("rgl")) install.packages("rgl", type="binary")
if (!require("RCurl")) install.packages("RCurl", type="binary")
if (!require("YuLab-SMU/ggmsa")) devtools::install_github("YuLab-SMU/ggmsa", type="binary")
if (!require("YuLab-SMU/treedataverse")) BiocManager::install("YuLab-SMU/treedataverse", type="binary")
if (!require("usedist")) devtools::install_github("kylebittinger/usedist")

##load required library
library(tidyverse)
library(phangorn)
library(ips)
library(msa)
library(bios2mds)
library(tinytex)
library(usedist)
library(tidyverse)
library(seqinr)
library(Biostrings)
library(microRNA)
library(ggtree)
library(ggimage)
library(ape)
library(ggtext)
library(ggmsa)
library(ggrepel)
#
#This will need to be set to your working directory
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
#knitr::opts_knit$set(root.dir = "~/Desktop/phiRes_workshop/workshop_data-analysis/phiRes_phylogeny")
```
```{r file read}
#SET input directory
indir<-"canidae/co1/fasta_files"
filelist<-list.files(indir, pattern = "fasta", full.names= TRUE)
seq<-Biostrings::readDNAStringSet(filelist)
seq<-RNA2DNA(seq)
```
```{r alignment}
##multi-sequence alignment.  Muscle is generally the fastest option here.  Other options are ClustalW and ClustalOmega

seqalign<-msa(seq, method= "Muscle", type = "dna", verbose = TRUE)
writeXStringSet(as(unmasked(seqalign), "XStringSet"), file="aln.fasta")
nt<-readDNAMultipleAlignment("aln.fasta", format = "fasta")

print(seqalign, show="complete")

##ggmsa for prettier alignments
ggmsa(nt, color = "Clustal", disagreement = TRUE, seq_name = TRUE, start = 100, end = 150) + 
  geom_seqlogo() + 
  geom_msaBar() 

pd<-msaConvert(seqalign,"phangorn::phyDat")
```
```{r Perform distance matrix and distance-based trees}

#margins on heatmap may need to be reduced.
dm <- dist.ml(pd)
heatmap(as.matrix(dm), margins=c(18,15), revC = TRUE,cexRow = 0.5, cexCol = 0.5, symm = TRUE)

treeUPGMA  <- upgma(dm)
plot(treeUPGMA, main="UPGMA")

treeNJ  <- NJ(dm)
plot(treeNJ, main="NJ")
```
```{r ML Model Selection}
##test models, select lowest AIC
mt <-modelTest(pd, multicore = TRUE, mc.cores= 6)
View(mt)
```
```{r Initial ModelFit}
##initial fit, use best fit model from prev chunk
fit <- as.pml(mt)
fit
```
```{r Optimize Fit and Bootstrap}
##MAKE SURE TO CHANGE OPTIONS DEPENDING ON YOUR SELECTED MODEL.  
#?optim.pml gives guidance on how to set optBf and optQ
#optInv and optGamma should be set depending on whether your model includes +I and/or +G parameters

fit.opt <- optim.pml(fit, optBf=TRUE, optQ=TRUE, optInv=TRUE, optGamma=TRUE,
                    rearrangement = "NNI", control = pml.control(trace = 0))
fit.opt
#main= sets the plot title
plot(fit.opt, main="TrN+I")

#bootstrap model
bs <- bootstrap.pml(fit.opt, bs=100, optNni=TRUE, multicore=TRUE, mc.cores = 6)
tree<-plotBS(fit.opt$tree, bs, p = 50, type="p")

#write tree with bootstrap node labels *Change name of output appropriately*
#write.tree(plotBS, file = "dog_CO1_ML_bs.newick")

```
```{r wrangle and attach metadata}

#note this may need to be changed depending on your data format
##important, for add_row line, (NA, X) needs to be modified for X so naming file has same number of records as data.  I don't know if this is a bug inherent to ggtree, but it is a workaround.

```
```{r generate trees}

#rooting functions

tree <- midpoint(tree)

#support for tip images

imgdir <- "canidae/pics"

#draw tree

#generate tip label database

tips <- as_tibble(tree$tip.label)

tips <- tips %>% 
  mutate(accession=str_extract(value,"[A-Z]{2}_?[0-9]{6}"),
         sciname=str_extract(value, "[A-Z]{1}[a-z]+ [a-z]+ ([a-z]+)?"),
         sciname=str_replace(sciname, "isolate", ""),
         sciname=str_replace(sciname, "from", ""),
         genus=str_extract(sciname, "^[A-Za-z]+ "),
         species=str_split_i(sciname, " ", 2),
         subspecies=str_split_i(sciname, " ", 3),
         isolate=str_extract(value, "(?<=isolate) .*"), 
         isolate=str_replace_all(isolate,"voucher.*","")
        ) 

treeplot <- ggtree(
  tree) %<+% tips +
  xlim(NA, 2) +
    geom_tiplab(
      aes(label=paste0('italic(', genus, ')~italic(', species, ')~italic(', subspecies, ')~', isolate, '')), parse=TRUE) +
  geom_nodelab(aes(label=label, 
      subset = !is.na(as.numeric(label)) & as.numeric(label) > 80)) +
  geom_tiplab(aes(subset=grepl("(Dane|Chihuahua)",label), image=paste0(imgdir, '/',label,'.jpg')), 
                geom="image", offset=1, align=2, size=0.1, by='height') +
  geom_label_repel(aes(label=label)) +
  geom_cladelab(node=1, label="test label", align=TRUE,  
                  offset = .2, textcolor='red', barcolor='red') +
  geom_treescale(x=0, y=45, width=1, color='red')
  
   treeplot   
  
   
imgdir <- "canidae/pics"
   
   
  (image=paste0("canidae/pics/",label,'.jpeg')), 
                geom="image", offset=2, align=2, size=.2)



```



